plugins/fpc/fu-fpc-device.c

```diff
@@ -0,0 +1,521 @@
+				   FALSE,
+				   error)) {
+		g_prefix_error(error, "failed to clear status: ");
+		return FALSE;
+	}
+
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_fpc_device_check_dfu_status,
+				  FPC_DFU_MAX_ATTEMPTS,
+				  20,
+				  (gpointer)&dfu_status,
+				  error))
+		return FALSE;
+
+	if (dfu_status.bmax_payload_size)
+		self->max_block_size = FPC_FLASH_BLOCK_SIZE_4096;
+	else
+		self->max_block_size = FPC_FLASH_BLOCK_SIZE_DEFAULT;
+
+	return TRUE;
+}
+
+static void
+fu_fpc_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+	fu_string_append_kx(str, idt, "Max block size", self->max_block_size);
+	fu_string_append_kb(str, idt, "Is a moc device", self->is_moc_device);
+}
+
+static gboolean
+fu_fpc_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	/* sanity check */
+	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in runtime mode, skipping");
+		return TRUE;
+	}
+	if (!fu_fpc_device_dfu_cmd(self, FPC_CMD_DFU_DETACH, 0x0000, NULL, 0, FALSE, FALSE, error))
+		return FALSE;
+	fu_device_remove_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_fpc_device_detach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	/* sanity check */
+	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in bootloader mode, skipping");
+		return TRUE;
+	}
+	if (!fu_fpc_device_fw_cmd(self, FPC_CMD_BOOT0, NULL, 0, FALSE, error))
+		return FALSE;
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	return TRUE;
+}
+
+static gboolean
+fu_fpc_device_setup(FuDevice *device, GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	/* setup */
+	if (!FU_DEVICE_CLASS(fu_fpc_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	self->is_moc_device = (fu_usb_device_get_pid(FU_USB_DEVICE(device)) >> 2 & 0x1) ? TRUE
+											: FALSE;
+
+	if (!fu_fpc_device_setup_mode(self, error)) {
+		g_prefix_error(error, "failed to get device mode: ");
+		return FALSE;
+	}
+
+	/* ensure version */
+	if (!fu_fpc_device_setup_version(self, error)) {
+		g_prefix_error(error, "failed to get firmware version: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_fpc_device_write_firmware(FuDevice *device,
+			     FuFirmware *firmware,
+			     FuProgress *progress,
+			     FwupdInstallFlags flags,
+			     GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+	FuFpcDfuStatus dfu_status = {0};
+	gboolean ret_dnload = TRUE;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
```

Comment:
if the values are the daemon-provided ones then you can remove this to get rid of the red warning.